OneSignal.js

```diff
@@ -64,6 +64,83 @@ var OneSignal = {
     xhr.open(action, ONESIGNAL_HOST_URL + url);
     xhr.setRequestHeader('Content-Type', 'application/json;charset=UTF-8');
     xhr.send(JSON.stringify(inData));
+  }
+};
+
+var __OSMessageHelper = {
+  // Dictionary of callbacks by type
+  // Each element is an array that can have a list of callbacks.
+  callbackHash: {},
+
+  isOnBackgroundPage: async function() {
+    return await new Promise(function(resolve, _reject) {
+      chrome.runtime.getBackgroundPage(function(page) {
+        resolve(page === window);
+      });
+    });
+  },
+
+  getOrCreateCallbackHashItem(type) {
+    const callbackList = __OSMessageHelper.callbackHash[type];
+    if (callbackList)
+      return callbackList;
+    __OSMessageHelper.callbackHash[type] = [];
+    return __OSMessageHelper.callbackHash[type];
+  },
+
+  // type - string callback name
+  // callback - function
+  addCallback: async function(type, callback) {
+    // Background page Context, we can directly fire the callback later when needed
+    if (await __OSMessageHelper.isOnBackgroundPage())
+      __OSMessageHelper.getOrCreateCallbackHashItem(type).push(callback);
+    else {
+      // Other context, setup a port to fire the callback as soon as it gets a message.
+      // chrome.runtime.onConnect will save from the other end.
+      const port = chrome.runtime.connect({name: type});
+      port.onMessage.addListener(function(msg) {
+        callback(msg);
+      });
+    }
+  },
+
+  setupBackgroundPagePortConnectListner: async function() {
+    if (!await __OSMessageHelper.isOnBackgroundPage())
+      return;
+    
+    chrome.runtime.onConnect.addListener(function(port) {
+      __OSMessageHelper.getOrCreateCallbackHashItem(port.name).push(port);
+    });
+  },
+
+  fireCallbacks: function(type, value) {
+    const callbackList = __OSMessageHelper.callbackHash[type];
+    if (!callbackList)
+      return;
+    callbackList.forEach(callback => {
+      if (__OSMessageHelper.isAPort(callback)) {
+        try {
+          callback.postMessage(value);
+        } catch(e) {
+          // Will throw if port is disconnected. Ignore this error.
+          // Port becomes disconnectd if a client page is closed.
+        }
+      }
+      else
+        callback(value);
+    });
+  },
+
+  isAPort: function(obj) {
```

Comment:
lets get rid of this (not needed because of new `portHash` object

Reply:
Fixed